homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,431 @@
+
+        if self._destination_entity_id is not None:
+            self._here_data.destination = await self._get_location_from_entity(
+                self._destination_entity_id)
+
+        await self._hass.async_add_executor_job(self._here_data.update)
+
+    async def _get_location_from_entity(self, entity_id: str) -> Optional[str]:
+        """Get the location from the entity state or attributes."""
+        entity = self._hass.states.get(entity_id)
+
+        if entity is None:
+            _LOGGER.error("Unable to find entity %s", entity_id)
+            return None
+
+        # Check if the entity has location attributes
+        if location.has_location(entity):
+            return self._get_location_from_attributes(entity)
+
+        # Check if device is in a zone
+        zone_entity = self._hass.states.get(
+            "zone.{}".format(entity.state))
+        if location.has_location(zone_entity):
+            _LOGGER.debug(
+                "%s is in %s, getting zone location",
+                entity_id, zone_entity.entity_id
+            )
+            return self._get_location_from_attributes(zone_entity)
+
+        # If zone was not found in state then use the state as the location
+        if entity_id.startswith("sensor."):
+            return entity.state
+
+        # When everything fails just return nothing
+        return None
+
+    @staticmethod
+    def _get_location_from_attributes(entity: State) -> str:
+        """Get the lat/long string from an entities attributes."""
+        attr = entity.attributes
+        return "{},{}".format(
+            attr.get(ATTR_LATITUDE), attr.get(ATTR_LONGITUDE)
+        )
+
+
+class HERETravelTimeData():
+    """HERETravelTime data object."""
+
+    def __init__(
+            self,
+            origin: None,
+            destination: None,
+            app_id: str,
+            app_code: str,
+            travel_mode: str,
+            traffic_mode: bool,
+            route_mode: str,
+            units: str) -> None:
+        """Initialize herepy."""
+        self.origin = origin
+        self.destination = destination
+        self.travel_mode = travel_mode
+        self.traffic_mode = traffic_mode
+        self.route_mode = route_mode
+        self.attribution = None
+        self.duration = None
+        self.distance = None
+        self.route = None
+        self.base_time = None
+        self.origin_name = None
+        self.destination_name = None
+        self.units = units
+        self._client = herepy.RoutingApi(app_id, app_code)
+
+    def update(self):
+        """Get the latest data from HERE."""
+        if self.traffic_mode:
+            traffic_mode = TRAFFIC_MODE_ENABLED
+        else:
+            traffic_mode = TRAFFIC_MODE_DISABLED
+
+        if self.destination is not None and self.origin is not None:
+            # Check for correct pattern
+            pattern = r"-?\d{1,2}\.\d+,-?\d{1,3}\.\d+"
+            if not re.fullmatch(pattern, self.origin):
```

Comment:
At this point it can be that the origin is from a resolved entity such as a template sensor. This sensor can hold the origin string which has to be checked here.

Reply:
Moreover a user can supply the friendly name of the zone. Config Schema validation cannot verify that this entity exists. As far as I see it the tests verify that all variations are tested here.

Reply:
Ok. Would it be better to have different options for different type of origins etc?

Reply:
Back when I first used `google_travel_time` and `waze_travel_time` components I was thrilled that I could just use the origin/destination field with so many options and didn't have to check the documentation all the time for special names and options.

Reply:
I'd like a second opinion from a member in this.

Reply:
My reason for implementing it: Before I implemented it, a false location could be send to the HERE API which then returned an error which was hard to understand.

If the current implementation is not wanted I could remove it, do the API request and if it fails log the error together with the origin/destination used.

Reply:
I'd like to know if we should continue this pattern of mixing different type of origins and destinations under the same config options. It impacts the complexity of the code, but might have user experience benefits, as mentioned above.

Reply:
Ideally I'd like to keep the functionality but in a base travel_time component, which google, waze,here,... make use of.

Reply:
Let's not do crazy regex but instead offer different config options. That makes it easier to validate and if things conflict, we don't have to make a decision what to pick.

Reply:
So we add configuration options `origin`, `origin_zone`, `origin_device_tracker` and so on?

Reply:
Have all origin options prefix with `origin_` and destination with `destination_`.

There seems to be these origin options in the waze integration:
origin_address
origin_coordinates
origin_entity_id
origin_zone

Please correct me as needed.

Reply:
I am not sure how to implement the `origin_coordinates`/`destination_coordinates`.
We do currently have a config validation for gps here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L68
But to use it the coordinates must be a tuple of floats not a string. I currently see 3 Options:

1. Use the existing gps validation and force the user to put the coordinates as a tuple of floats (`90.0,-90.0` instead of `"90.0,-90.0"`). This is completely different from the other travel_time options and might confuse users. It would also need many changes to the current code.

2. Use `origin_latitude`/`origin_longitude` with their respective [validators](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L62-L67). Both can be float or string as they are coerced to float by voluptuous. This again is different from the other implementations and might confuse users. It would only need a few changes to the current code.

3. Use the current form of `"90.0,-90.0"` and add a new validator `gps_string`. The configuration would be the same as now and no code changes would be required.

Reply:
Please use number 2.

Reply:
Is implemented